modules/rapid/src/histogram.cpp

```diff
@@ -0,0 +1,226 @@
+
+            // threshold
+            if(dst.at<uchar>(i, j) < threshold)
+                dst.at<uchar>(i, j) = 0;
+        }
+    }
+}
+
+static void calcHueSatHist(const Mat_<Vec3b>& hsv, Mat_<float>& hist)
+{
+    for (int i = 0; i < hsv.rows; i++)
+    {
+        for (int j = 0; j < hsv.cols; j++)
+        {
+            const Vec3b& c = hsv(i, j);
+            // thresholds as in sec. 4.1
+            if (c[1] > 25 && c[2] > 50)
+            {
+                hist(c[0] * hist.rows / 256, c[1] * hist.cols / 256)++;
+            }
+        }
+    }
+}
+
+static float sum(Mat_<float>& hist)
+{
+    CV_DbgAssert(hist.isContinuous());
+    float ret = 0;
+    int N = int(hist.total());
+    const float* ptr = hist.ptr<float>();
+    for (int i = 0; i < N; i++)
+        ret += ptr[i];
+    return ret;
+}
+
+static double bhattacharyya_coeff(const Mat& a, const Mat& b)
+{
+    CV_DbgAssert(a.isContinuous() && b.isContinuous());
+    int N = int(a.total());
+    double ret = 0;
+    const float* aptr = a.ptr<float>();
+    const float* bptr = b.ptr<float>();
+    for (int i = 0; i < N; i++)
+        ret += std::sqrt(aptr[i] * bptr[i]);
+    return ret;
+}
+
+static void findCorrespondenciesOLS(const cv::Mat_<float>& scores, cv::Mat_<int>& cols)
+{
+    cols.resize(scores.rows);
+    for (int i = 0; i < scores.rows; i++)
+    {
+        int pos = -1;
+        for (int j = scores.cols - 1; j >= 0; j--)
+        {
+            if (scores(i, j) >= 0.35)
+            {
+                pos = j;
+                break;
+            }
+        }
+
+        cols(i) = pos;
+    }
+}
+
+struct OLSTrackerImpl : public OLSTracker
+{
+    Mat vtx;
+    Mat tris;
+
+    Mat_<float> fgHist;
+    Mat_<float> bgHist;
+    double tau;
+    uchar sobelThresh;
+
+    OLSTrackerImpl(InputArray _pts3d, InputArray _tris, int histBins, uchar _sobelThesh)
+    {
+        CV_Assert(_tris.getMat().checkVector(3, CV_32S) > 0);
+        CV_Assert(_pts3d.getMat().checkVector(3, CV_32F) > 0);
+        vtx = _pts3d.getMat();
+        tris = _tris.getMat();
+
+        tau = 1.0; // currently does not work as intended. effectively disable
+        sobelThresh = _sobelThesh;
+
+        bgHist.create(histBins, histBins);
+    }
+
+    void computeAppearanceScore(const Mat& bundleHSV, const Mat& bundleGrad, Mat_<float>& score) const
+    {
+        score.resize(bundleHSV.rows);
+        score = 0;
+        Mat_<float> hist(fgHist.size());
+
+        for (int i = 0; i < bundleHSV.rows; i++)
+        {
+            int start = 0;
+            for (int j = 0; j < bundleHSV.cols; j++)
+            {
+                if (bundleGrad.at<uchar>(i, j))
+                {
+                    // compute the histogram between last candidate point to current candidate point
+                    // as in eq. (4)
+                    hist = 0;
+                    calcHueSatHist(bundleHSV({i, i + 1}, {start, j}), hist);
+                    hist /= std::max(sum(hist), 1.0f);
```

Comment:
```
float histSum = cv::sum(hist)[0];
hist /= std::max(histSum, 1.0f);
```
? (Pseudocode)

Arguably a little more complicated here, but saves the custom `sum`-function.

Reply:
the custom sum function is due to performance reasons as the dispatch code in cv::sum becomes relevant when you are summing 8x8 px.